src/devices/sound/k054539.cpp

```diff
@@ -104,13 +104,34 @@ void k054539_device::keyoff(int channel)
 		regs[0x22c] &= ~(1 << channel);
 }
 
+void k054539_device::advance_filter(int channel, int val)
+{
+	float *hist = filter_hist[channel];
+	hist[0] = hist[1];
+	hist[1] = hist[2];
+	hist[2] = hist[3];
+	hist[3] = ((float) val) / 32768.0f;
+}
+
+float k054539_device::calculate_filter(int channel, float t)
+{
+	// Cubic hermite interpolation
+	// t is domain [0,1]
+
+	const float *hist = filter_hist[channel];
+	const float a = (-hist[0] / 2.0f) + (3.0f * hist[1] / 2.0f) - (3.0f * hist[2] / 2.0f) + (hist[3] / 2.0f);
+	const float b = hist[0] - (5.0f * hist[1] / 2.0f) + (2.0f * hist[2]) - (hist[3] / 2.0f);
+	const float c = (-hist[0] / 2.0f) + (hist[2] / 2.0f);
+	return (a * t * t * t) + (b * t * t) + (c * t) + hist[1];
```

Comment:
Due to floating point error accumulation, the compiler can’t lift the common multiplications out of the expression here.  However, if you express it as `(t * t * t * a) + (t * t * b) + (t * c) + hist[1]` the compiler *can* lift the common `t * t`.

You could also manually lift it like:
```
   const float t2 = t * t;
   return (a * t2 * t) + (b * t2) + (c * t) + hist[1];
```
But that’s a bit more verbose.